juju/juju
# [JUJU-1778] Simplify db-accessor worker with OptionFactory usage

Here we add to the Dqlite `OptionFactory`, the remaining option emissions required for clustering:
- `WithTLS`
- `WithCluster`

We also pass a logger into the factory and proxy Dqlite logs through it via `WithLogFunc`.

The option factory is then utilised by the db-accessor worker instead of the worker-specific logic to recruit Dqlite options.

At this time we take a similar posture to the `peergrouper` worker in that we require controller nodes to have a unique local-cloud address for Dqlite to bind to. What we do not yet accommodate is a configured value for `juju-ha-space`. It also has to be noted that the current approach whereby we look at agent configuration for API addresses means that these are affected by a configured `juju-ctrl-space` value, which may not be desirable.

## QA steps

- `make go-install`
- Bootstrap to LXD.
- `juju enable-ha`.
- Watch status for the controller model until quiescence.
- Check `juju debug-log -m controller --replay|grep dbaccessor|less` and observe logged cluster comms:
```
machine-1: 11:29:16 INFO juju.worker.dbaccessor initialized dqlite application (ID: 12428324535580141454)
machine-2: 11:29:19 DEBUG juju.worker.dbaccessor determined Dqlite bind address: 10.161.87.97
machine-2: 11:29:19 DEBUG juju.worker.dbaccessor determined Dqlite cluster members: [10.161.87.222:17666]
machine-0: 11:29:19 DEBUG juju.worker.dbaccessor new connection from 10.161.87.97:51026
machine-2: 11:29:19 DEBUG juju.worker.dbaccessor attempt 1: server 10.161.87.222:17666: connected
machine-2: 11:29:19 DEBUG juju.worker.dbaccessor new connection from 10.161.87.97:54756
machine-1: 11:29:19 DEBUG juju.worker.dbaccessor new connection from 10.161.87.97:33908
machine-0: 11:29:19 DEBUG juju.worker.dbaccessor new connection from 10.161.87.97:51032
machine-2: 11:29:19 DEBUG juju.worker.dbaccessor new connection from 10.161.87.222:34030
machine-0: 11:29:19 DEBUG juju.worker.dbaccessor new connection from 10.161.87.97:51036
machine-1: 11:29:19 DEBUG juju.worker.dbaccessor new connection from 10.161.87.222:53464
machine-0: 11:29:19 DEBUG juju.worker.dbaccessor new connection from 10.161.87.43:40026
machine-2: 11:29:20 INFO juju.worker.dbaccessor serving REPL over "/var/lib/juju/dqlite/juju.sock"
machine-2: 11:29:20 INFO juju.worker.dbaccessor initialized dqlite application (ID: 17679387133952718839)
machine-0: 11:29:26 DEBUG juju.worker.dbaccessor new connection from 10.161.87.222:60526
machine-0: 11:29:26 DEBUG juju.worker.dbaccessor attempt 1: server 10.161.87.222:17666: connected
```

## Documentation changes

None.

## Bug reference

N/A
